lib/balanced-pool.js

```diff
@@ -73,20 +123,42 @@ class BalancedPool extends PoolBase {
       throw new BalancedPoolMissingUpstreamError()
     }
 
-    let dispatcher = this[kClients].find(dispatcher => (
-      !dispatcher[kNeedDrain] &&
-      dispatcher.closed !== true &&
-      dispatcher.destroyed !== true
-    ))
+    const allClientsBusy = this[kClients].map(pool => pool[kNeedDrain]).reduce((a, b) => a && b, true)
 
-    if (!dispatcher) {
+    if (allClientsBusy) {
       return
     }
 
-    this[kClients].splice(this[kClients].indexOf(dispatcher), 1)
-    this[kClients].push(dispatcher)
+    let counter = 0
+
```

Comment:
I think you just want the non busy clients for the rest of the logic here, i.e:

```
const clients = this[kClients].filter(pool => !pool[kNeedDrain])
```

Reply:
i'll loose the offset and mapping of kIndex will be wrong